File: transfer codes/hauffman.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
	char symbol;
	int freq;
	Node *left, *right;
	Node(char symbol, int freq)
	{
		this->symbol = symbol;
		this->freq = freq;
		left = right = nullptr;
	}
};

struct Compare
{
	bool operator()(Node *const &n1, Node *const &n2)
	{
		return n1->freq > n2->freq;
	}
};

void util(Node *root, string code)
{
	if (!root)
		return;

	if (!root->left && !root->right)
	{
		cout << root->symbol << " -> " << code << endl;
	}

	util(root->left, code + "0");
	util(root->right, code + "1");
}

int main()
{
	priority_queue<Node *, vector<Node *>, Compare> pq;
	int T;
	cin >> T;

	while (T--)
	{
		char c;
		int f;
		cin >> c >> f;
		Node *cr = new Node(c, f);
		pq.push(cr);
	}

	while (pq.size() > 1)
	{
		Node *p1 = pq.top();
		pq.pop();
		Node *p2 = pq.top();
		pq.pop();

		Node *t = new Node('$', p1->freq + p2->freq);
		t->left = p1;
		t->right = p2;
		pq.push(t);
	}

	Node *root = pq.top();

	util(root, "");

	return 0;
}

File: transfer codes/hauffman.py
import heapq

class Node:
	def __init__(self, symbol, freq):
		self.symbol = symbol
		self.freq = freq
		self.right = None
		self.left = None
	# comparator for heap 
	def __lt__(self, other):
		return self.freq < other.freq

def printHuffman(root, codeStr):
	if root is None:
		return
	
	# if it is leaf node 
	if root.left is None and root.right is None:
		print(f"{root.symbol} -> {codeStr}")

	printHuffman(root.left, codeStr+"0")
	printHuffman(root.right, codeStr+"1")

def buildTree(inp):
	heap = []

	for symbol, freq in inp:
		heapq.heappush(heap, Node(symbol, freq))

	while len(heap) > 1:
		left = heapq.heappop(heap)
		right = heapq.heappop(heap)

		parent = Node('$', left.freq + right.freq)
		parent.left = left
		parent.right = right

		heapq.heappush(heap, parent)
	
	# return the laste remaining node (root of the our huffman tree)
	return heap[0]

def main():
	T = int(input())

	# list to take input
	inp = []

	for i in range(T):
		symbol = input()
		freq = int(input())
		inp.append((symbol, freq))

	root = buildTree(inp)

	print("The hauffman for each letter is : ")
	printHuffman(root, "") # before root string will be empty

if __name__ == "__main__":
	main()